File: problem=023/Solution.java
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

final class Solution {

	private static final boolean isAbundant(int n) {
		final int limit = (int) Math.sqrt(n);
		int sum = 0;
		for (int i = 1; i <= limit && sum <= n; i++) {
			if (n % i == 0) {
				sum += i;
				if (i != 1 && n / i != i) {
					sum += n / i;
				}
			}
		}
		return sum > n;
	}

	private static final boolean canBeExpressedAsSumOfTwoAbundantNumbers(int n, List<Integer> list) {
		for (final int anAbundantNumber : list) {
			final int counterNumber = n - anAbundantNumber;
			if (Collections.binarySearch(list, counterNumber) >= 0) {
				return true;
			}
		}
		return false;
	}

	private static long solve(int limit) {
		final List<Integer> abundantList = new ArrayList<>(limit);
		for (int i = 1; i <= limit; i++) {
			if (isAbundant(i)) {
				abundantList.add(i);
			}
		}
		long sum = 0;
		for (int i = 1; i <= limit; i++) {
			if (!canBeExpressedAsSumOfTwoAbundantNumbers(i, abundantList)) {
				sum += i;
			}

		}
		return sum;
	}

	public static void main(String[] args) throws Exception {
		System.out.println(solve(28123));
	}
}

// If this solution helped you, please star the repo!

File: problem=023/Solution.js
const isAbundant = n => {
	const limit = Math.floor(Math.sqrt(n));
	let sum = 0;
	for (let i = 1; i <= limit && sum <= n; i++) {
		if (n % i == 0) {
			sum += i;
			if (i != 1 && n / i != i) {
				sum += n / i;
			}
		}
	}
	return sum > n;
};

const binarySearch = (list, n) => {
	let left = 0,
		right = list.length - 1;
	while (left <= right) {
		const mid = Math.floor((left + right) / 2);
		if (n === list[mid]) {
			return mid;
		}
		if (n < list[mid]) {
			right = mid - 1;
		} else {
			left = mid + 1;
		}
	}
	return -1;
};

const abundantList = [];

const canBeExpressedAsSumOfTwoAbundantNumbers = n => {
	for (let i = 0; i < abundantList.length; i++) {
		const anAbundantNumber = abundantList[i];
		const counterNumber = n - anAbundantNumber;
		if (binarySearch(abundantList, counterNumber) >= 0) {
			return true;
		}
	}
	return false;
};

const solve = limit => {
	for (let i = 1; i <= limit; i++) {
		if (isAbundant(i)) {
			abundantList.push(i);
		}
	}
	let sum = 0;
	for (let i = 1; i <= limit; i++) {
		if (!canBeExpressedAsSumOfTwoAbundantNumbers(i)) {
			sum += i;
		}
	}
	return sum;
};

console.log(solve(28123));

// If this solution helped you, please star the repo!
